### PR TITLE
Introduce conv3d_transpose in tf.layers

### LINKED ISSUE #8359: tf.layers.conv3d_transpose missing
In tf.layers, there are conv2d and conv2d_transpose. There is conv3d, but no conv3d_transpose. 

I'm wondering if it is possible to have conv3d_transpose any time soon. 

Thanks,

### PR DESCRIPTION
**Pull request by kdexd**
This PR attempts to introduce `Conv3DTranspose` class and an equivalent functional interface named `conv3d_transpose`. Additions have been done in **tensorflow/python/layers/convolution.py**

Inclusion consists of additions of class, method, documentation, tests and aliases.

Fixes #8359

### CODE REVIEW

**Comment by karandesai-96** [comment]
I signed it!

**Comment by karandesai-96** [comment]
@tensorflow-jenkins test this please

**Comment by karandesai-96** [comment]
I have completed my part of work, ready for reviews. So shall we start Jenkins builds ? 
/ping: @jhseu (checked using `git blame`).

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by karandesai-96** [comment]
Does a member need to always write "Jenkins, test this please" to run the tests ? I made a minor typo, just fixed it.

**Comment by lukaszkaiser** [comment]
Sorry, I'm not too fluent in conv layers and have never used the 3d ones. Unassigning myself.

**Comment by karandesai-96** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Jenkins, test this please

**Comment by jhseu** [comment]
Francois, mind reviewing this pull request?

**Comment by fchollet** [comment]
I'm on vacation for a few days. I'll get to it, but it will take a while.

On Mar 17, 2017 11:22 AM, "Jonathan Hseu" <notifications@github.com> wrote:

Francois, mind reviewing this pull request?

—
You are receiving this because your review was requested.

Reply to this email directly, view it on GitHub
<https://github.com/tensorflow/tensorflow/pull/8461#issuecomment-287433408>,
or mute the thread

File: tensorflow/python/layers/convolutional.py [hunk 1]
```diff
@@ -1200,16 +1201,238 @@ def conv2d_transpose(inputs,
   return layer.apply(inputs)
 
 
+class Conv3DTranspose(Conv3D):
+  """Transposed 3D convolution layer (sometimes called Deconvolution).
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: "2D deconvolution"

File: tensorflow/python/layers/convolutional.py [hunk 2]
```diff
@@ -1200,16 +1201,238 @@ def conv2d_transpose(inputs,
...
+
+  The need for transposed convolutions generally arises
+  from the desire to use a transformation going in the opposite direction
+  of a normal convolution, i.e., from something that has the shape of the
+  output of some convolution to something that has the shape of its input
+  while maintaining a connectivity pattern that is compatible with
+  said convolution.
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: This seems redundant with the docstring for `Conv2DTranspose`. No need to have it twice.

File: tensorflow/python/layers/convolutional.py [hunk 3]
```diff
@@ -1200,16 +1201,238 @@ def conv2d_transpose(inputs,
...
+
+  def call(self, inputs):
+    inputs_shape = array_ops.shape(inputs)
+    batch_size = inputs_shape[0]
+    if self.data_format == 'channels_first':
+      c_axis, d_axis, h_axis, w_axis = 1, 2, 3, 4
+    else:
+      c_axis, d_axis, h_axis, w_axis = 4, 1, 2, 3
+
+    depth, height, width = inputs_shape[d_axis], inputs_shape[h_axis], \
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: Avoid this line break pattern. Split this into several lines.

File: tensorflow/python/layers/convolutional.py [hunk 4]
```diff
@@ -1200,16 +1201,238 @@ def conv2d_transpose(inputs,
...
+        inputs_shape[w_axis]
+    kernel_d, kernel_h, kernel_w = self.kernel_size
+    stride_d, stride_h, stride_w = self.strides
+
+    def get_deconv_dim(dim_size, stride_size, kernel_size, padding):
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: We may want to put this in `utils`. Adding it here would mean we have it twice in the codebase.

File: tensorflow/python/layers/convolutional.py [hunk 5]
```diff
@@ -1200,16 +1201,238 @@ def conv2d_transpose(inputs,
...
+    out_shape[d_axis] = get_deconv_dim(
+        out_shape[d_axis], stride_d, kernel_d, self.padding)
+    out_shape[h_axis] = get_deconv_dim(
+        out_shape[h_axis], stride_h, kernel_h, self.padding)
+    out_shape[w_axis] = get_deconv_dim(
+        out_shape[w_axis], stride_w, kernel_w, self.padding)
+    outputs.set_shape(out_shape)
+
+    if self.bias:
+      outputs = nn.bias_add(
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: I believe this would fail in data_format `channels_first` (to be verified). See what we do for `Conv3D` bias.

File: tensorflow/python/layers/convolutional.py (hunk 1 above)
Comment (karandesai-96) [inline tensorflow/python/layers/convolutional.py]: I am sorry I didn't get you, did you mean 2D or 3D ?

Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: 3D, sorry

**Comment by karandesai-96** [comment]
@fchollet: Well your reviews were a little unusual for me to interpret, because certain chunks of code were modified on upstream master after I submitted my patch, and before you reviewed my PR. I rebased my branch on master and worked on your reviews.

I had a difficulty in setting up development mode of TF, but Bazel is all up and running on my local machine. All tests pass properly locally, pleas initiate Jenkins builds and have a look at the modifications, thanks !

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@fchollet please take a look at new set of commits. Thanks!

**Comment by karandesai-96** [comment]
@drpngx I made a very silly error while resolving conflicts during rebasing. Probably made a wrong push, sorry to bother you. Please help me in initiating the builds, they should pass now.

File: tensorflow/python/layers/convolutional.py [hunk 6]
```diff
@@ -1149,7 +1139,8 @@ def conv2d_transpose(inputs,
                      trainable=True,
                      name=None,
                      reuse=None):
-  """Transposed convolution layer (sometimes called Deconvolution).
+  """Functional interface for transposed 2D convolution layer (sometimes
+  called 2D Deconvolution).
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: The one-line docstring summary should fit on one line. No need to mention "sometimes called 2D Deconvolution".

File: tensorflow/python/layers/convolutional.py [hunk 7]
```diff
@@ -1212,16 +1203,260 @@ def conv2d_transpose(inputs,
...
+        bias_regularizer=bias_regularizer,
+        activity_regularizer=activity_regularizer,
+        trainable=trainable,
+        name=name, **kwargs)
+
+  def build(self, input_shape):
+    if len(input_shape) != 5:
+      raise ValueError('Inputs should have rank ' +
+                       str(5) +
+                       'Received input shape:', str(input_shape))
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: Please add punctuation. This prints "Inputs should have rank 5Received"

File: tensorflow/python/layers/convolutional.py [hunk 8]
```diff
@@ -1212,16 +1203,260 @@ def conv2d_transpose(inputs,
...
+        kernel_regularizer=kernel_regularizer,
+        bias_regularizer=bias_regularizer,
+        activity_regularizer=activity_regularizer,
+        trainable=trainable,
+        name=name, **kwargs)
+
+  def build(self, input_shape):
+    if len(input_shape) != 5:
+      raise ValueError('Inputs should have rank ' +
+                       str(5) +
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: No need for `str`. This should be part of the error message string

File: tensorflow/python/layers/convolutional.py [hunk 9]
```diff
@@ -1212,16 +1203,260 @@ def conv2d_transpose(inputs,
...
+                     kernel_initializer=None,
+                     bias_initializer=init_ops.zeros_initializer(),
+                     kernel_regularizer=None,
+                     bias_regularizer=None,
+                     activity_regularizer=None,
+                     trainable=True,
+                     name=None,
+                     reuse=None):
+  """Functional interface for transposed 3D convolution layer (sometimes
+  called 3D Deconvolution).
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: Make it fit on one line

File: tensorflow/python/layers/utils.py [hunk 10]
```diff
@@ -164,3 +165,29 @@ def constant_value(pred):
   else:
     raise TypeError('`pred` must be a Tensor, a Variable, or a Python bool.')
   return pred_value
+
+
+def get_deconv_dim(dim_size, stride_size, kernel_size, padding):
+  """Return the output dimension of a particular deconvolution layer, given
+  the input dimensions.
```
Comment (fchollet) [inline tensorflow/python/layers/utils.py]: Make it fit on one line

File: tensorflow/python/layers/convolutional.py [hunk 11]
```diff
@@ -1212,16 +1203,260 @@ def conv2d_transpose(inputs,
...
+      raise ValueError('Inputs should have rank ' +
+                       str(5) +
+                       'Received input shape:', str(input_shape))
+    if self.data_format == 'channels_first':
+      channel_axis = 1
+    else:
+      channel_axis = -1
+    if input_shape[channel_axis] is None:
+      raise ValueError('The channel dimension of the inputs '
+                       'should be defined. Found `None`.')
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: Printing the full input shape would be helpful here.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by karandesai-96** [comment]
Finally Jenkins goes green ! :checkered_flag: Making the final set of changes now.

**Comment by drpngx** [comment]
Woohoo!

**Comment by karandesai-96** [comment]
Worked on second set of reviews by @fchollet.

**Review by fchollet — APPROVED** [review]
It looks good to me; since this is a large change, anyone else would like to give it a second good?

**Comment by karandesai-96** [comment]
Thanks @fchollet. Also may I ask if support of `channels_first` for `nn.bias_add` is absent due to a particular reason or you are open for contributions ? The codebase is huge but I can give it a try.

**Comment by drpngx** [comment]
Jenkins, test this please.

I'll take a look.

**Review by drpngx — CHANGES_REQUESTED** [review]

File: tensorflow/python/layers/convolutional.py [hunk 12]
```diff
@@ -1212,16 +1202,258 @@ def conv2d_transpose(inputs,
   return layer.apply(inputs)
 
 
+class Conv3DTranspose(Conv3D):
+  """Transposed 3D convolution layer (sometimes called 3D Deconvolution).
+
+  Arguments:
+    filters: Integer, the dimensionality of the output space (i.e. the number
```
Comment (fchollet) [inline tensorflow/python/layers/convolutional.py]: This is for consistency with the convolutional layer API in general.

File: tensorflow/python/layers/utils.py [hunk 13]
```diff
@@ -164,3 +165,28 @@ def constant_value(pred):
...
+
+  Returns:
+    An int representing the size of output dimension of the layer.
+  """
+  if isinstance(dim_size, ops.Tensor):
+    dim_size = math_ops.multiply(dim_size, stride_size)
+  elif dim_size is not None:
+    dim_size *= stride_size
+
+  if padding == 'valid' and dim_size is not None:
```
Comment (fchollet) [inline tensorflow/python/layers/utils.py]: It gets normalized to lowercase in the constructor.

File: tensorflow/python/ops/nn_ops.py [hunk 14]
```diff
@@ -1293,8 +1293,7 @@ def conv3d_transpose(value,
       dimension of the input tensor.
     padding: A string, either `'VALID'` or `'SAME'`. The padding algorithm.
       See the @{tf.nn.convolution$comment here}
-    data_format: A string, either `'NDHWC'` or `'NCDHW`' specifying the layout
-      of the input and output tensors. Defaults to `'NDHWC'`.
+    data_format: A string. 'NDHWC' and 'NCDHW' are supported.
```
Comment (karandesai-96) [inline tensorflow/python/ops/nn_ops.py]: I will have to just remove "defaults to ndhwc" as default is None.

**Review by drpngx — APPROVED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Just to be clear, this does not expose anything until you modify the [`tf.layers` interface](https://www.github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/layers.py#L52). In that case, it would need an API review.

**Comment by karandesai-96** [comment]
_ping_

**Comment by drpngx** [comment]
We still need an API review. @martinwicke @jhseu perhaps?

**Comment by fchollet** [comment]
@drpngx the API is good.

**Comment by karandesai-96** [comment]
Okay so if everything is well, then i shall expose this layer in functional API.

**Comment by jhseu** [comment]
Added this PR to the TF API review list for their meeting next week.

**Comment by karandesai-96** [comment]
Okay so if everything is well, then i shall expose this layer in functional API.
Also, is it required to manually make a RELEASE.md entry about a new layer interface or does some bot do it ?

**Comment by jhseu** [comment]
It's manual. Feel free to add an entry in this PR.

**Comment by drpngx** [comment]
@karandesai-96 feel free to push a new commit with the release notes then let me know.

@fchollet thanks!

**Comment by karandesai-96** [comment]
@drpngx I made a changelog entry and as the master had moved forward a lot, I did a rebase of this branch to make sure that my code is compatible with current master.

**Comment by drpngx** [comment]
Thanks!

Jenkins, test this please.

**Comment by karandesai-96** [comment]
A test related to XLA is failing on MacOS CPU. I doubt whether it is related to this PR. I am having a Linux machine, I could neither understand it by looking on Jenkins console, nor have resources to reproduce the bug. How should I proceed ?

**Comment by karandesai-96** [comment]
@drpngx I just checked on master, there were a couple of commits which have these failures, but it was rectified a few hours ago. I guess if I rebase again, this shall be solved. Rebasing to master again !

**Comment by karandesai-96** [comment]
Done ! Thanks for too many "Jenkins, test this please" @drpngx. I believe I would let you do that less frequently from my next PR, got to know a lot of things about Tensorflow dev from my first one :smile:

**Comment by drpngx** [comment]
Jenkins, test this please.

We'll have to plant a lot of trees to offset this PR ;-)

File: tensorflow/python/ops/nn_ops.py [hunk 15]
```diff
@@ -1294,7 +1294,7 @@ def conv3d_transpose(value,
     padding: A string, either `'VALID'` or `'SAME'`. The padding algorithm.
       See the @{tf.nn.convolution$comment here}
     data_format: A string, either `'NDHWC'` or `'NCDHW`' specifying the layout
-      of the input and output tensors. Defaults to `'NDHWC'`.
```
Comment (martinwicke) [inline tensorflow/python/ops/nn_ops.py]: Does it not have a default any more?

**Review by martinwicke — CHANGES_REQUESTED** [review]
This is fine from an API POV. Left one comment about the docstring.

File: tensorflow/python/ops/nn_ops.py (hunk 15 above)
Comment (karandesai-96) [inline tensorflow/python/ops/nn_ops.py]: @martinwicke I did a rebase on master at some point of time, data_format is `None` by default, just like other method signatures. But the thing is, it acts like `NDHWC` being a NoneType, if not `NCDHW`. I modified the docstring because the default value is not `NDHWC` explicitly.

Comment (karandesai-96) [inline tensorflow/python/ops/nn_ops.py]: Either we should keep this docstring or we should modify the method signatures of all these ops to have a default `data_format`, which is a string instead of NoneType. And probably introduce some flexibility with `channels_first`, `channels_last` and all these `NHWC` formats. Currently passed `NHWC` to `tf.layers` constructors throws an error which I personally found uncomfortable.

Your opinions @fchollet @drpngx ?

Comment (fchollet) [inline tensorflow/python/ops/nn_ops.py]: >  And probably introduce some flexibility with channels_first, channels_last and all these NHWC formats.

We will do that, this is planned.

> Either we should keep this docstring or we should modify the method signatures of all these ops to have a default data_format, which is a string instead of NoneType. 

Having a string as default would be preferable, *unless* we want to make the default globally configurable (change it once and have it apply to all ops for which no data format is specified).

Comment (karandesai-96) [inline tensorflow/python/ops/nn_ops.py]: If we are making the default globally configurable then it would be ambiguous to specify strings in method signatures, might mislead users visiting the API. In that scenario, we can have NoneType.

All cases considered, in my humble opinion, to match current method signatures of other ops in this script, we should go with NoneType at least for this PR. Hence the phrase "Defaults to NDHWC" wasn't valid in the docstring because that's not the default value even if the op implicitly behaves in the same manner when that value would have been NDHWC.

Rest I would leave the decision on veterans and make modifications accordingly.

CC: @martinwicke

Comment (martinwicke) [inline tensorflow/python/ops/nn_ops.py]: Let's add the default explicitly. We cannot change it anyway because that would break compatibility, so we might as well encode it formally.

Comment (karandesai-96) [inline tensorflow/python/ops/nn_ops.py]: `max_pool`, `avg_pool`, `conv2d_tranpose` and many other ops have `data_format="NHWC"` specified in function signature. I've done the same for this for the sake of consistency.

/ping: @martinwicke :rocket:

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Hooray!

**Comment by karandesai-96** [comment]
Thanks @martinwicke @fchollet @drpngx, first contribution to tensorflow! Feels good to contribute back to the community :smile:

**Comment by drpngx** [comment]
Thank you and congratulations!
